Batch of ngtsc changes for Material, CLI, g3 compatibility.

alxhub opened the pull request.

alxhub committed:
feat(ivy): resolve forwardRef() for queries

@ContentChild[ren] and @ViewChild[ren] can contain a forwardRef() to a
type. This commit allows ngtsc to unwrap the forward reference and
deal with the node inside.

It includes two modes of support for forward reference resolution -
a foreign function resolver which understands deeply nested forward
references in expressions that are being statically evaluated, and
an unwrapForwardRef() function which deals only with top-level nodes.

Both will be useful in the future, but for now only unwrapForwardRef()
is used.

alxhub committed:
fix(ivy): force new imports for .d.ts files

When ngtsc encounters a reference to a type (for example, a Component
type listed in an NgModule declarations array), it traces the import
of that type and attempts to determine the best way to refer to it.

In the event the type is defined in the same file where a reference
is being generated, the identifier of the type is used. If the type
was imported, ngtsc has a choice. It can use the identifier from the
original import, or it can write a new import to the module where the
type came from.

ngtsc has a bug currently when it elects to rely on the user's import.
When writing a .d.ts file, the user's import may have been elided as
the type was not referred to from the type side of the program. Thus,
in .d.ts files ngtsc must always assume the import may not exist, and
generate a new one.

In .js output the import is guaranteed to still exist, so it's
preferable for ngtsc to continue using the existing import if one is
available.

This commit changes how @angular/compiler writes type definitions, and
allows it to use a different expression to write a type definition than
is used to write the value. This allows ngtsc to specify that types in
type definitions should always be imported. A corresponding change to
the staticallyResolve() Reference system allows the choice of which
type of import to use when generating an Expression from a Reference.

gkalpak commented on packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts:
```diff
@@ -253,7 +254,9 @@ const UNARY_OPERATORS = new Map<ts.SyntaxKind, (a: any) => any>([
-  foreignFunctionResolver?(node: ts.FunctionDeclaration|ts.MethodDeclaration): ts.Expression|null;
+  foreignFunctionResolver?
+      (ref: Reference<ts.FunctionDeclaration|ts.MethodDeclaration>,
+       args: ReadonlyArray<ts.Expression>): ts.Expression|null;
```
Why a `ReadonlyArray` here but not in other places?

alxhub replied:
So it can directly accept the `.arguments` property of a `ts.CallExpression`, which is a `NodeArray` that extends `ReadonlyArray`.

gkalpak commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -103,3 +103,33 @@ export function unwrapExpression(node: ts.Expression): ts.Expression {
+    return node;
+  }
+  const arg = node.arguments[0];
+  if (!ts.isArrowFunction(arg) || ts.isBlock(arg.body)) {
```
Why does it have to be an arrow function?
Wouldn't that be a problem for (f)esm5 in ngcc?

alxhub replied:
Good point. Fixed.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -14,7 +14,7 @@ import {Reference, filterToMembersWithDecorator, reflectObjectLiteral, staticall
 
 import {SelectorScopeRegistry} from './selector_scope';
-import {getConstructorDependencies, isAngularCore, unwrapExpression} from './util';
+import {forwardRefResolver, getConstructorDependencies, isAngularCore, unwrapExpression, unwrapForwardRef} from './util';
```
`forwardRefResolver` looks unused, could it be removed

alxhub replied:
Done.

vicb commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -103,3 +103,33 @@ export function unwrapExpression(node: ts.Expression): ts.Expression {
   return node;
 }
+
+export function unwrapForwardRef(node: ts.Expression, reflector: ReflectionHost): ts.Expression {
```
Please add a one liner comment

alxhub replied:
Done.

vicb commented on packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts:
```diff
@@ -78,6 +78,11 @@ export interface ResolvedValueArray extends Array<ResolvedValue> {}
 type Scope = Map<ts.ParameterDeclaration, ResolvedValue>;
 
+export enum ImportMode {
+  USE_EXISTING_IMPORT,
```
use const enum ?

alxhub replied:
Done. No need for `const enum`, this never ends up in the browser.

vicb reviewed (dismissed):
I added a few minor comments. LGTM otherwise

alxhub committed:
fix(ivy): correctly write cross-file references

There is a bug in the existing handling for cross-file references.
Suppose there are two files, module.ts and component.ts.

component.ts declares two components, one of which uses the other.
In the Ivy model, this means the component will get a directives:
reference to the other in its defineComponent call.

That reference is generated by looking at the declared components
of the module (in module.ts). However, the way ngtsc tracks this
reference, it ends up comparing the identifier of the component
in module.ts with the component.ts file, detecting they're not in
the same file, and generating a relative import.

This commit changes ngtsc to track all identifiers of a reference,
including the one by which it is declared. This allows toExpression()
to correctly decide that a local reference is okay in component.ts.

mary-poppins commented:
You can preview 6d97ae0 at https://pr25080-6d97ae0.ngbuilds.io/.

mary-poppins commented:
You can preview ea60fe3 at https://pr25080-ea60fe3.ngbuilds.io/.

alxhub committed:
fix(ivy): allow relative imports of .d.ts files

ngtsc used to assume that all .d.ts dependencies (that is, third party
packages) were imported via an absolute module path. It turns out this
assumption isn't valid; some build tools allow relative imports of
other compilation units.

In the absolute case, ngtsc assumes (and still does) that all referenced
types are available through the entrypoint from which an @NgModule was
imported. This commit adds support for relative imports, in which case
ngtsc will use relative path resolution to determine the imports.

alxhub committed:
fix(ivy): don't use a custom ts.CompilerHost for ngtsc

ngtsc used to have a custom ts.CompilerHost which delegated to the plain
ts.CompilerHost. There's no need for this wrapper class and it causes
issues with CLI integration, so delete it.

alxhub committed:
fix(bazel): allow compile_strategy to be (privately) imported

compile_strategy() is used to decide whether to build Angular code
using ngc (legacy) or ngtsc (local). In order for g3 BUILD rules
to switch properly and allow testing of Ivy in g3, they need to
import this function.

This commit removes the _ prefix which allows the function to be
imported.

alxhub committed:
fix(ivy): wait for preanalyze promises in loadNgStructureAsync()

loadNgStructureAsync() for ngtsc has a bug where it returns a
Promise<Promise[]> instead of awaiting the entire array of Promises.

This commit uses Promise.all() to await the whole set.

alxhub committed:
fix(ivy): don't crash in listLazyRoutes()

This commit replaces the "not implemented" error when calling
listLazyRoutes() with an empty result, which will allow testing
in the CLI before listLazyRoutes() is implemented.

mary-poppins commented:
You can preview 054df5e at https://pr25080-054df5e.ngbuilds.io/.

alxhub commented:
http://test/OCL:206011842:BASE:206227538:1532642558270:eca7b697

mhevery reviewed (approved):
Looks reasonable to me. Please take care of comments by others before merging.

alfaproject commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -75,12 +75,11 @@ export class NgtscProgram implements api.Program {
-          .map(file => this.compilation !.analyzeAsync(file))
-          .filter((result): result is Promise<void> => result !== undefined);
     }
+    await Promise.all(this.tsProgram.getSourceFiles()
```
Why not just return the promise and get rid of the `async` keyword? Seems like you are using async/await just for the sake of it o:
